README.rst
=====
kwsgi
=====

Yet another WSGI server implementation.


Usage
-----

Usage is like this:

.. code-block:: console

   $ kwsgi hello.py app --reload

Options are below:

.. code-block:: console

   $ kwsgi --help
   Usage: kwsgi [OPTIONS] FILEPATH WSGIAPP

     Example: kwsgi hello.py app -p 5000 --reload

   Options:
     -h, --host TEXT               The interface to bind to.
     -p, --port INTEGER            The port to bind to.
     --reload / --no-reload        Enable live reloading
     --daemonize / --no-daemonize  Detaches the server from the controlling
                                   terminal and enters the background.
     --interval INTEGER            Interval time to check file changed for
                                   reloading
     --validate / --no-validate    Validating your WSGI application complying with
                                   PEP3333 compliance.
     --help                        Show this message and exit.


And you can integrate with kwsgi from python script:

.. code-block:: python

   from kwsgi import WSGIServer


   def application(env, start_response):
       start_response('200 OK', [('Content-type', 'text/plain; charset=utf-8')])
       return [b'Hello World']


   if __name__ == '__main__':
       server = WSGIServer(application)
       server.run_forever()


Development Roadmap
-------------------

These are the current planned major milestones:

1. [DONE] Add minimum implementation (at least this can run django application).
2. Optimize some important performance bottlenecks using C-extensions, mypyc or Rust.
3. Add green threads implementation.


CHANGES.rst
=======
CHANGES
=======

0.0.2 (2015-09-29)
------------------

* Add daemonize option.

0.0.1 (2015-09-25)
------------------

* Create this project.


kwsgi/server.py
import socket
import urllib.parse
from threading import Thread


class ResponseWriter:
    def __init__(self):
        self.headers = None
        self.status_code = None

    def start_response(self, status_code, headers):
        self.status_code = status_code
        self.headers = headers

    @property
    def called(self):
        return self.headers is not None and self.status_code is not None


def worker(conn, wsgi_app, env):
    with conn:
        response = ResponseWriter()
        wsgi_response = wsgi_app(env, response.start_response)
        print(f"{env['REMOTE_ADDR']} - {response.status_code}")

        if not response.called:
            conn.sendall(b'HTTP/1.1 500\r\n\r\nInternal Server Error\n')
            return

        status_line = f"HTTP/1.1 {response.status_code}".encode("utf-8")
        headers = [f"{k}: {v}" for k, v in response.headers]

        response_body = b""
        content_length = 0
        for b in wsgi_response:
            response_body += b
            content_length += len(b)
        headers.append(f"Content-Length: {content_length}")
        header_bytes = "\r\n".join(headers).encode("utf-8")
        env["wsgi.input"].close()  # this doesn't raise error if close twice.
        conn.sendall(status_line + b"\r\n" + header_bytes + b"\r\n\r\n" + response_body)


class WSGIServer:
    def __init__(self, app, host="127.0.0.1", port=8000,
                 max_accept=128, timeout=30.0, rbufsize=-1):
        self.app = app
        self.host = host
        self.port = port
        self.max_accept = max_accept
        self.timeout = timeout
        self.rbufsize = rbufsize

    def make_wsgi_environ(self, rfile, client_address):
        # should return '414 URI Too Long' if line is longer than 65536.
        raw_request_line = rfile.readline(65537)
        method, path, version = str(raw_request_line, 'iso-8859-1').rstrip('\r\n').split(' ', maxsplit=2)
        if '?' in path:
            path, query = path.split('?', 1)
        else:
            path, query = path, ''

        env = {
            'REQUEST_METHOD': method,
            'PATH_INFO': urllib.parse.unquote(path, 'iso-8859-1'),
            'QUERY_STRING': query,
            'SERVER_PROTOCOL': "HTTP/1.1",
            'SERVER_NAME': socket.getfqdn(),
            'SERVER_PORT': self.port,
            'REMOTE_ADDR': client_address[0],
            'SCRIPT_NAME': "",
            'wsgi.version': (1, 0),
            'wsgi.url_scheme': "http",
            'wsgi.multithread': True,
            'wsgi.multiprocess': False,
            'wsgi.run_once': False,
        }

        while True:
            # should return '431 Request Header Fields Too Large'
            # if line is longer than 65536 or header exceeds 100 lines.
            line = rfile.readline(65537)
            if line in (b'\r\n', b'\n', b''):
                break

            key, value = line.decode('iso-8859-1').rstrip("\r\n").split(":", maxsplit=1)
            value = value.lstrip(" ")
            if key.upper() == "CONTENT-TYPE":
                env['CONTENT_TYPE'] = value
            if key.upper() == "CONTENT-LENGTH":
                env['CONTENT_LENGTH'] = value
            env_key = "HTTP_" + key.replace("-", "_").upper()
            if env_key in env:
                env[env_key] = env[env_key] + ',' + value
            else:
                env[env_key] = value
        env['wsgi.input'] = rfile
        return env

    def run_forever(self):
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, True)
            sock.bind((self.host, self.port))
            sock.listen(self.max_accept)

            while True:
                conn, client_address = sock.accept()
                conn.settimeout(self.timeout)
                rfile = conn.makefile('rb', self.rbufsize)
                env = self.make_wsgi_environ(rfile, client_address)
                Thread(target=worker, args=(conn, self.app, env), daemon=True).start()


kwsgi/reloader.py
import os
import sys
import threading
import time
import _thread
import subprocess
import tempfile


# For reloading server when detected python files changes.
EXIT_STATUS_RELOAD = 3


class FileCheckerThread(threading.Thread):
    """ Interrupt main-thread as soon as a changed module file is detected,
        the lockfile gets deleted or gets too old. """
    # This class is copied and pasted from following source code of Bottle (MIT License).
    # https://github.com/bottlepy/bottle/blob/5aeb2562de4c8133dcba8a8cf49b6ec90cd15d1f/bottle.py#L3732-L3771

    def __init__(self, lockfile, interval):
        threading.Thread.__init__(self)
        self.daemon = True
        self.lockfile, self.interval = lockfile, interval
        #: Is one of 'reload', 'error' or 'exit'
        self.status = None

    def run(self):
        files = dict()

        for module in list(sys.modules.values()):
            path = getattr(module, '__file__', '')
            if path[-4:] in ('.pyo', '.pyc'):
                path = path[:-1]
            if path and os.path.exists(path):
                files[path] = os.stat(path).st_mtime

        while not self.status:
            if not os.path.exists(self.lockfile) or \
                    os.stat(self.lockfile).st_mtime < time.time() - self.interval - 5:
                self.status = 'error'
                _thread.interrupt_main()
            for path, last_mtime in files.items():
                if not os.path.exists(path) or os.stat(path).st_mtime > last_mtime:
                    self.status = 'reload'
                    _thread.interrupt_main()
                    break
            time.sleep(self.interval)

    def __enter__(self):
        self.start()

    def __exit__(self, exc_type, *_):
        if not self.status:
            self.status = 'exit'  # silent exit
        self.join()
        return exc_type is not None and issubclass(exc_type, KeyboardInterrupt)


class AutoReloadServer:
    def __init__(self, func, args=None, kwargs=None):
        self.func = func
        self.func_args = args
        self.func_kwargs = kwargs

    def run_forever(self, interval):
        if not os.environ.get('KWSGI_CHILD'):
            lockfile = None
            try:
                fd, lockfile = tempfile.mkstemp(prefix='kwsgi.', suffix='.lock')
                os.close(fd)  # We only need this file to exist. We never write to it
                while os.path.exists(lockfile):
                    args = [sys.executable] + sys.argv
                    environ = os.environ.copy()
                    environ['KWSGI_CHILD'] = 'true'
                    environ['KWSGI_LOCKFILE'] = lockfile
                    p = subprocess.Popen(args, env=environ)
                    while p.poll() is None:  # Busy wait...
                        os.utime(lockfile, None)  # Alive! If lockfile is unlinked, it raises FileNotFoundError.
                        time.sleep(interval)
                    if p.poll() != EXIT_STATUS_RELOAD:
                        if os.path.exists(lockfile):
                            os.unlink(lockfile)
                            sys.exit(p.poll())
            except KeyboardInterrupt:
                pass
            finally:
                if os.path.exists(lockfile):
                    os.unlink(lockfile)
            return

        try:
            lockfile = os.environ.get('KWSGI_LOCKFILE')
            bgcheck = FileCheckerThread(lockfile, interval)
            with bgcheck:
                self.func(*self.func_args, **self.func_kwargs)
            if bgcheck.status == 'reload':
                sys.exit(EXIT_STATUS_RELOAD)
        except KeyboardInterrupt:
            pass
        except (SystemExit, MemoryError):
            raise
        except:
            time.sleep(interval)
            sys.exit(EXIT_STATUS_RELOAD)


kwsgi/daemonize.py
import os
import sys


DEVNULL = getattr(os, 'devnull', '/dev/null')


def daemonize():
    """Detaches the server from the controlling terminal and enters the background."""
    if 'KWSGI_CHILD' in os.environ:
        return

    if os.fork():
        # Exit a parent process because setsid() will be fail
        # if you're a process group leader.
        sys.exit(0)

    # Detaches the process from the controlling terminal.
    os.setsid()

    if os.fork():
        sys.exit(0)

    # Continue to run application if directory is removed.
    os.chdir('/')

    # 0o22 means 755 (777-755)
    os.umask(0o22)

    # Remap all of stdin, stdout and stderr on to /dev/null.
    # Please caution that this way couldn't support following execution:
    #   $ kwsgi ... > output.log 2>&1
    os.closerange(0, 3)
    fd_null = os.open(DEVNULL, os.O_RDWR)
    if fd_null != 0:
        os.dup2(fd_null, 0)
    os.dup2(fd_null, 1)
    os.dup2(fd_null, 2)


kwsgi/cli.py
import os
import sys
from importlib.machinery import SourceFileLoader

import click

from .reloader import AutoReloadServer
from .server import WSGIServer
from .daemonize import daemonize as daemonize_func


def insert_import_path_to_sys_modules(import_path):
    """
    When importing a module, Python references the directories in sys.path.
    The default value of sys.path varies depending on the system, But:
    When you start Python with a script, the directory of the script is inserted into sys.path[0].
    So we have to replace sys.path to import object in specified scripts.
    """
    abspath = os.path.abspath(import_path)
    if os.path.isdir(abspath):
        sys.path.insert(0, abspath)
    else:
        sys.path.insert(0, os.path.dirname(abspath))


def run_server(app, host, port):
    click.echo('Start: {host}:{port}'.format(host=host, port=port))
    server = WSGIServer(app, host=host, port=port)
    server.run_forever()


def run_live_reloading_server(interval, app, host, port):
    click.echo('Start: {host}:{port}'.format(host=host, port=port))
    server = AutoReloadServer(run_server, kwargs={'app': app, 'host': host, 'port': port})
    server.run_forever(interval)


@click.command()
@click.argument('filepath', nargs=1, envvar='KWSGI_FILE', type=click.Path(exists=True))
@click.argument('wsgiapp', nargs=1, envvar='KWSGI_WSGI_APP')
@click.option('--host', '-h', type=click.STRING, default='127.0.0.1', envvar='KWSGI_HOST',
              help='The interface to bind to.')
@click.option('--port', '-p', type=click.INT, default=8000, envvar='KWSGI_PORT',
              help='The port to bind to.')
@click.option('--reload/--no-reload', default=False, envvar='KWSGI_RELOAD',
              help='Enable live reloading')
@click.option('--daemonize/--no-daemonize', default=False, envvar='KWSGI_DAEMONIZE',
              help='Detaches the server from the controlling terminal and enters the background.')
@click.option('--interval', type=click.INT, default=1, envvar='KWSGI_INTERVAL',
              help='Interval time to check file changed for reloading')
@click.option('--validate/--no-validate', default=False, envvar='KWSGI_VALIDATE',
              help='Validating your WSGI application complying with PEP3333 compliance.')
def cli(filepath, wsgiapp, host, port, reload, daemonize, interval, validate):
    """
    Example: kwsgi hello.py app -p 5000 --reload
    """
    insert_import_path_to_sys_modules(filepath)
    module = SourceFileLoader('module', filepath).load_module()
    app = getattr(module, wsgiapp)

    if validate:
        from wsgiref.validate import validator
        app = validator(app)

    if reload and daemonize:
        click.echo("You couldn't use the both of --reload and --daemonize.")
        click.echo("Because when enabling daemonize option, called chdir system call"
                   " to continue to run application if the directory is removed.")
        sys.exit(1)

    if daemonize:
        daemonize_func()
        run_server(app=app, host=host, port=port)

    if reload:
        run_live_reloading_server(interval, app=app, host=host, port=port)
    else:
        run_server(app=app, host=host, port=port)
